presto-main/src/main/java/com/facebook/presto/cost/CostComparator.java

```diff
@@ -58,6 +62,19 @@ public int compare(Session session, PlanNodeCostEstimate left, PlanNodeCostEstim
         requireNonNull(left, "left can not be null");
         requireNonNull(right, "right can not be null");
         checkArgument(!left.isUnknown() && !right.isUnknown(), "cannot compare unknown costs");
+
+        long maxQueryMemoryPerNode = nodeMemoryConfig.getMaxQueryMemoryPerNode().toBytes();
+        long maxQueryMemory = SystemSessionProperties.getQueryMaxMemory(session).toBytes();
+
+        long memoryLimit = maxQueryMemory > maxQueryMemoryPerNode ?  maxQueryMemoryPerNode : maxQueryMemory;
```

Comment:
for maxQueryMemoryPerNode is it possible to check that` queryMemory/numberOfNodes < maxQueryMemoryPerNode`?

so then it would looks something like

```
if(exceedsMemoryLimit(left.getMemoryCost()) && !exceedsMemoryLimit(right.getMemoryCost()) {
    return -1;
}
...
private boolean exceedsMemoryLimit(long memoryCost) {
 // do the memory checks here
}
```